### PR TITLE
[don't merge] Implement [LCC] Carmen, Cruel Skymarcher

### PR DESCRIPTION
**Pull request by skaspels**
Added and tested, funcitoning correctly.

### CODE REVIEW

File: Mage.Sets/src/mage/cards/c/CarmenCruelSkymarcher.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+        // Flying
+        this.addAbility(FlyingAbility.getInstance());
+
+        // Whenever a player sacrifices a permanent, put a +1/+1 counter on Carmen, Cruel Skymarcher and you gain 1 life.
+        this.addAbility(new CarmenTriggeredAbility());
+
+        // Whenever Carmen attacks, return up to one target permanent card with mana value less than or equal to Carmen's power from your graveyard to the battlefield.
+        Ability ability = new AttacksTriggeredAbility(new ReturnFromGraveyardToBattlefieldTargetEffect());
+        ability.addTarget(new TargetCardInYourGraveyard(filter));
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/c/CarmenCruelSkymarcher.java:54]: Miss up to one settings (0, 1)?

File: Mage.Sets/src/mage/cards/c/CarmenCruelSkymarcher.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+/**
+ *
+ * @author skaspels
+ */
+public final class CarmenCruelSkymarcher extends CardImpl {
+
+    private static final FilterCard filter = new FilterPermanentCard(
+            " a permanent from your graveyard with mana value less than or equal to {this}'s power");
+    static {
+        filter.add(CarmenPredicate.instance);
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/c/CarmenCruelSkymarcher.java:35]: Miss card owner filter (from your graveyard)

Comment (xenohedron) [reply Mage.Sets/src/mage/cards/c/CarmenCruelSkymarcher.java:35]: Doesn't the target class handle that?

Comment (JayDi85) [reply Mage.Sets/src/mage/cards/c/CarmenCruelSkymarcher.java:35]: Yes, fine

Comment (skaspels) [reply Mage.Sets/src/mage/cards/c/CarmenCruelSkymarcher.java:35]: The ability does hand that - `ability.addTarget(new TargetCardInYourGraveyard(filter));`
will add the min number of targets shortly.

**Review by skaspels — COMMENTED** [review]
This is the corrected filter

**Comment by skaspels** [comment]
It seems that each of my committs is rolled up into this one pull request - do I need to make a separate [ull request for the implement [lcc] contest of claws card? I would think so, but it doesn't seem to me that I can create a new pull request for it; github seems to treat it as part of the previous pull request i opened. When i try to open the pull request from magefree/master, I can never find my commits or my branch on the branch selection box.

File: Mage.Sets/src/mage/cards/c/CarmenCruelSkymarcher.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        super(card);
+    }
+
+    @Override
+    public CarmenCruelSkymarcher copy() {
+        return new CarmenCruelSkymarcher(this);
+    }
+}
+
+class CarmenTriggeredAbility extends TriggeredAbilityImpl {
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/c/CarmenCruelSkymarcher.java:68]: this is okay as is for now but I'm making a note to clean up this and other cards to common classes (we have SacrificeAllTriggeredAbility and SacrificePermanentTriggeredAbility)

File: Mage.Sets/src/mage/cards/c/CarmenCruelSkymarcher.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    public CarmenTriggeredAbility copy() {
+        return new CarmenTriggeredAbility(this);
+    }
+
+    @Override
+    public String getRule() {
+        return "Whenever a player sacrifices a permanent, put a +1/+1 counter on {this} and you gain 1 life.";
+    }
+}
+enum CarmenPredicate implements ObjectSourcePlayerPredicate<Card> {
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/c/CarmenCruelSkymarcher.java:99]: I should also look into making this a common predicate, it's used in a few places

File: Mage.Sets/src/mage/cards/c/CarmenCruelSkymarcher.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    @Override
+    public CarmenTriggeredAbility copy() {
+        return new CarmenTriggeredAbility(this);
+    }
+
+    @Override
+    public String getRule() {
+        return "Whenever a player sacrifices a permanent, put a +1/+1 counter on {this} and you gain 1 life.";
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/c/CarmenCruelSkymarcher.java:96]: fyi: avoid overriding getRule. It's better to generate as much text automatically as possible... can help catch mistakes if the text doesn't come out as expected. In this case the effects should come out fine (just with `.concatBy("and")` on the second) so would just need to `setTriggerPhrase` in the constructor. But don't worry about it for this PR.

File: Mage.Sets/src/mage/cards/c/ContestOfClaws.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        if (myPermanent == null) {
+            return false;
+        }
+        int damageValue = myPermanent.getPower().getValue();
+        Permanent anotherPermanent = game.getPermanent(source.getTargets().get(1).getFirstTarget());
+        if (anotherPermanent == null){
+            return false;
+        }
+        int lethal = Math.min(anotherPermanent.getLethalDamage(source.getSourceId(), game), damageValue);
+        anotherPermanent.damage(damageValue, source.getSourceId(), source, game);
```
Comment (xenohedron) [inline Mage.Sets/src/mage/cards/c/ContestOfClaws.java:79]: this looks wrong. the source of the damage is the first target creature, not the spell.

**Review by xenohedron — COMMENTED** [review]
In future, you need to commit to separate branches. (Committing to your master branch and then merging upstream into it is already starting to get messy.) Combining multiple implementations into one PR slows down the review and merge process.

If you can, please rebase / cherry-pick these two implementations each onto a fresh branch without extraneous commits. There are many tutorials for how to use git/github on the web.

**Comment by xenohedron** [comment]
maybe one of these screenshots will help

image

image

**Comment by xenohedron** [comment]
Alright I cleaned up your master branch so it's in sync. From now on make your commits to a separate branch.

**Pull request closed without merging**